### PR TITLE
Based rematch

### PR DESCRIPTION
**Pull request by mekealbrown**

### CODE REVIEW

File: src/based-rematch.cpp [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+*  8-23 -> G16
+*  24-59 -> M38
+*/
+static const std::vector<std::string> base_nums = {"!", "@", "#", "$", "%", "^", "&", "*", "{", "}", "[", "]", \
+    "(", ")", "`", ">", "a", "b", "c", "d", "e", "f", "g", "h", "|", "~", ",", "/", "?", "_", "+", "=", "-", \
+    ":", "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"};
+
+
+//set indices into base_nums vector
+void set_start_end(int base, int *start, int *end)
```
Comment (malloc-nbytes) [inline src/based-rematch.cpp]: use references

File: src/based-rematch.cpp [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+      *end = S8_END;
+      break;
+    case 16:
+      *start = G16_START;
+      *end = G16_END;
+      break;
+    case 36:
+      *start = M36_START;
+      *end = M36_END;
+      break;
```
Comment (malloc-nbytes) [inline src/based-rematch.cpp]: make a default case

File: src/based-rematch.cpp [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+  }
+}
+
+//return the index of the symbol in relation to its base from base_nums
+int find_val(int base, std::string symbol)
```
Comment (malloc-nbytes) [inline src/based-rematch.cpp]: make symbol reference

File: src/based-rematch.cpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+{
+  int start, end;
+  set_start_end(base, &start, &end);
+  int i = 0;
+  for(; start <= end; start++) {
```
Comment (malloc-nbytes) [inline src/based-rematch.cpp]: make for loop: `for (int i = 0; start <= end; start++)`

File: src/based-rematch.cpp [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    if(base_nums[start] == symbol) {
+      return i;
+    }
+    ++i;
+  }
+  std::cout << "didnt find value...\n\n"; //shouldn't ever get here ofc
+  return -1;
```
Comment (malloc-nbytes) [inline src/based-rematch.cpp]: probably should crash here or have the caller crash if it receives -1

File: src/based-rematch.cpp [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+//convert each value to base 10, puts it in "solved" vector
+void values_solve(strvec_t vals, std::vector<int> bases, std::vector<int> *solved)
```
Comment (malloc-nbytes) [inline src/based-rematch.cpp]: use reference instead of pointer

File: src/based-rematch.cpp [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    {
+      case 0:
+        bases.push_back(8);
+        break;
+      case 1:
+        bases.push_back(16);
+        break;
+      case 2:
+        bases.push_back(36);
+        break;
```
Comment (malloc-nbytes) [inline src/based-rematch.cpp]: add default case

File: src/based-rematch.cpp [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+  for (const auto& pair : paired_vec) {
+      values.push_back(pair.second);
+      //std::cout << pair.second << "\n"; //for debug purposes
+  }
+  //std::cout << "\n";
+
+  int key = utils_rng_roll(10000, 950000, seed);
+  values.push_back(std::to_string(key));
+
+  FLAGS |= (NO_HDR | NO_FTR);
```
Comment (malloc-nbytes) [inline src/based-rematch.cpp]: Turn this off when done with it

File: src/rematch-based-puzzle.cpp [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+*  8-23 -> G16
+*  24-59 -> M38
+*/
+static const std::vector<std::string> base_nums = {"!", "@", "#", "$", "%", "^", "&", "*", "{", "}", "[", "]", \
+    "(", ")", "`", ">", "a", "b", "c", "d", "e", "f", "g", "h", "|", "~", ",", "/", "?", "_", "+", "=", "-", \
+    ":", "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"};
+
+
+//set indices into base_nums vector
+void set_start_end(int base, int *start, int *end)
```
Comment (malloc-nbytes) [inline src/rematch-based-puzzle.cpp:37]: use references

File: src/rematch-based-puzzle.cpp [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+      *end = S8_END;
+      break;
+    case 16:
+      *start = G16_START;
+      *end = G16_END;
+      break;
+    case 36:
+      *start = M36_START;
+      *end = M36_END;
+      break;
```
Comment (malloc-nbytes) [inline src/rematch-based-puzzle.cpp:52]: add default

File: src/rematch-based-puzzle.cpp [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+  }
+}
+
+//return the index of the symbol in relation to its base from base_nums
+int find_val(int base, std::string symbol)
```
Comment (malloc-nbytes) [inline src/rematch-based-puzzle.cpp:57]: use reference

File: src/rematch-based-puzzle.cpp [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+{
+  int start, end;
+  set_start_end(base, &start, &end);
+  int i = 0;
```
Comment (malloc-nbytes) [inline src/rematch-based-puzzle.cpp:61]: put `i` in the for loop initialization

File: src/rematch-based-puzzle.cpp [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+  for(; start <= end; start++) {
+    if(base_nums[start] == symbol) {
+      return i;
+    }
+    ++i;
+  }
+  std::cout << "didnt find value...\n\n"; //shouldn't ever get here ofc
+  return -1;
```
Comment (malloc-nbytes) [inline src/rematch-based-puzzle.cpp:69]: have this crash or have caller crash if received -1

File: src/rematch-based-puzzle.cpp [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+//convert each value to base 10, puts it in "solved" vector
+void values_solve(strvec_t vals, std::vector<int> bases, std::vector<int> *solved)
```
Comment (malloc-nbytes) [inline src/rematch-based-puzzle.cpp:73]: use reference

File: src/rematch-based-puzzle.cpp [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+    {
+      case 0:
+        bases.push_back(8);
+        break;
+      case 1:
+        bases.push_back(16);
+        break;
+      case 2:
+        bases.push_back(36);
+        break;
```
Comment (malloc-nbytes) [inline src/rematch-based-puzzle.cpp:123]: add default

File: src/rematch-based-puzzle.cpp [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+  for (const auto& pair : paired_vec) {
+      values.push_back(pair.second);
+      //std::cout << pair.second << "\n"; //for debug purposes
+  }
+  //std::cout << "\n";
+
+  int key = utils_rng_roll(10000, 950000, seed);
+  values.push_back(std::to_string(key));
+
+  FLAGS |= (NO_HDR | NO_FTR);
```
Comment (malloc-nbytes) [inline src/rematch-based-puzzle.cpp:144]: disable these flags when done with them

**Review by malloc-nbytes — CHANGES_REQUESTED** [review]

File: src/based-rematch.cpp (hunk 4 above)
Comment (mekealbrown) [reply src/based-rematch.cpp]: done

File: src/based-rematch.cpp (hunk 3 above)
Comment (mekealbrown) [reply src/based-rematch.cpp]: each string is only 4 characters, im gonna leave it as by-value

File: src/based-rematch.cpp (hunk 5 above)
Comment (mekealbrown) [reply src/based-rematch.cpp]: fixed

File: src/rematch-based-puzzle.cpp (hunk 12 above)
Comment (mekealbrown) [reply src/rematch-based-puzzle.cpp]: fixed

File: src/rematch-based-puzzle.cpp (hunk 13 above)
Comment (mekealbrown) [reply src/rematch-based-puzzle.cpp:68]: fixed

File: src/rematch-based-puzzle.cpp (hunk 14 above)
Comment (mekealbrown) [reply src/rematch-based-puzzle.cpp:72]: fixed

File: src/rematch-based-puzzle.cpp (hunk 15 above)
Comment (mekealbrown) [reply src/rematch-based-puzzle.cpp:122]: fixed

**Pull request merged**